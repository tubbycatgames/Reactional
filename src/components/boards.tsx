import React from 'react'
import {
  Directions,
  FlingGestureHandler,
  State,
} from 'react-native-gesture-handler'

type Props = {
  // TODO Hack to get over the reducer having an odd type
  children: any
  onDown?: Function
  onLeft?: Function
  onRight?: Function
  onUp?: Function
}

export const SwipeBoard = ({
  children,
  onDown,
  onLeft,
  onRight,
  onUp,
}: Props) => {
  const swipeEvents: [Directions, Function?][] = [
    [Directions.DOWN, onDown],
    [Directions.LEFT, onLeft],
    [Directions.RIGHT, onRight],
    [Directions.UP, onUp],
  ]
  return swipeEvents.reduce(
    (tree, [direction, fn]) =>
      fn ? (
        <FlingGestureHandler
          direction={direction}
          onHandlerStateChange={({ nativeEvent }) => {
            if (nativeEvent.state === State.ACTIVE) {
              fn()
            }
          }}
        >
          {tree}
        </FlingGestureHandler>
      ) : (
        tree
      ),
    children
  )
}
