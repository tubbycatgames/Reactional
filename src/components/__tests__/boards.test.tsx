import React from 'react'
import renderer from 'react-test-renderer'
import { Text } from 'react-native'

import { SwipeBoard } from '../boards'
import { State } from 'react-native-gesture-handler'

let handlerTag = 1

describe('SwipeBoard', () => {
  it('should render children when no events passed', () => {
    const render = renderer.create(
      <SwipeBoard>
        <Text>Plain</Text>
      </SwipeBoard>
    )
    expect(render.toJSON()).toMatchSnapshot()
  })

  it('should render wrapped children when an event passed', () => {
    const event = jest.fn()
    const render = renderer.create(
      <SwipeBoard onDown={event}>
        <Text>Plain</Text>
      </SwipeBoard>
    )

    const text = render.root.findByType(Text)
    text.props.onGestureHandlerStateChange({
      nativeEvent: { handlerTag, state: State.ACTIVE },
    })
    handlerTag++
    expect(event).toHaveBeenCalledTimes(1)

    expect(render.toJSON()).toMatchSnapshot()
  })

  it('should wrap children up to the number of possible events', () => {
    const event = jest.fn()
    const render = renderer.create(
      <SwipeBoard onDown={event} onLeft={event} onRight={event} onUp={event}>
        <Text>Plain</Text>
      </SwipeBoard>
    )

    const expectedCalls = 4
    const goalTag = handlerTag + expectedCalls
    const text = render.root.findByType(Text)
    while (handlerTag <= goalTag) {
      text.props.onGestureHandlerStateChange({
        nativeEvent: { handlerTag, state: State.ACTIVE },
      })
      handlerTag++
    }
    expect(event).toHaveBeenCalledTimes(expectedCalls)
  })
})
