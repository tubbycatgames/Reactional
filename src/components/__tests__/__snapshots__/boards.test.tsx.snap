// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeBoard should render children when no events passed 1`] = `
<Text>
  Plain
</Text>
`;

exports[`SwipeBoard should render wrapped children when an event passed 1`] = `
<Text
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
>
  Plain
</Text>
`;
